### PR TITLE
Metrics/refactor small improvements

### PR DESCRIPTION
**Pull request by ph**
This PR contains:
- a few speed improvements for collecting metrics by removing unnecessary allocations.
- A way to get the uptime for a pipeline and the agent
- a filterdelegator to collect filter stats.

### CODE REVIEW

**Comment by ph** [comment]
This PR also add a method a way to format the duration correctly for a human.

`Pipeline#uptime` => time in milliseconds since started
`Agent#uptime` => Time in milliseconds since started

You can have a human string with `LogStash::FormatDuration.format(milliseconds)` this will return something like this `4d 45m`

File: logstash-core/spec/logstash/instrument/null_metric_spec.rb [hunk 1]
```diff
@@ -7,4 +7,16 @@
   it "defines the same interface as `Metric`" do
     expect(described_class).to implement_interface_of(LogStash::Instrument::Metric) 
   end
+
+  describe "#time" do
+    it "returns the vaule of the block without recording any metrics" do
+      x = 1
```
Comment (purbon) [inline logstash-core/spec/logstash/instrument/null_metric_spec.rb]: why this variable here?

File: logstash-core/spec/logstash/output_delegator_spec.rb [hunk 2]
```diff
@@ -15,10 +15,20 @@
 
     before do
       allow(out_klass).to receive(:new).with(any_args).and_return(out_inst)
+      allow(out_inst).to receive(:register)
+      allow(out_inst).to receive(:identifier_name)
+      allow(out_inst).to receive(:metric=).with(any_args)
+      allow(logger).to receive(:debug).with(any_args)
+    end
+
+    before do
```
Comment (purbon) [inline logstash-core/spec/logstash/output_delegator_spec.rb]: I would advise to use always ```before(:each)```  or ```before(:all)```, as relying on the default it gets me always confused for the expected behaviour.

File: logstash-core/lib/logstash/webserver.rb [hunk 3]
```diff
@@ -5,8 +5,7 @@
 require 'puma/configuration'
 require 'puma/commonlogger'
 
-module LogStash
-  class WebServer
+module LogStash class WebServer
```
Comment (purbon) [inline logstash-core/lib/logstash/webserver.rb]: why this change? it looks to me a bit cosmetic, before it was actually more clear to understand, right?

**Comment by purbon** [comment]
@ph can we add also here a method in the agent that lets query the pipelines? I envision the API holding a reference to the agent and then this being responsible to ask the pipelines. What do you think?

**Comment by purbon** [comment]
@ph can you rebase this branch, files like ```info_command.rb``` were actually removed. This change should happen in the new command used for the root resource.

File: logstash-core/spec/logstash/instrument/null_metric_spec.rb (hunk 1 above)
Comment (ph) [inline logstash-core/spec/logstash/instrument/null_metric_spec.rb]: you are right I could replace that with a string.

**Comment by ph** [comment]
@purbon rebased, cleaned with your comments.

added a command in agent to retrieve the pipelines.

**Comment by purbon** [comment]
@ph I guess this need now another rebase?

File: logstash-core/lib/logstash/instrument/metric_store.rb [hunk 4]
```diff
@@ -37,8 +39,8 @@ def get(*key_paths)
     # Take all the individuals `MetricType` and convert them to `Logstash::Event`
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: This documentation should be updated.

File: logstash-core/lib/logstash/instrument/metric_store.rb [hunk 5]
```diff
@@ -37,8 +39,8 @@ def get(*key_paths)
...
     #
     # @return [Array] An array of all metric transformed in `Logstash::Event`
-    def to_events
-      to_events_recursively(@store).flatten
+    def all_metrics
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: not totally comfortable with a name like ```all_metrics``` for a class named ```MetricStore```, what do you think? might be named it simply ```all``` ?

File: logstash-core/lib/logstash/instrument/null_metric.rb [hunk 6]
```diff
@@ -1,6 +1,16 @@
...
+require "logstash/instrument/metric"
+
 module LogStash module Instrument
  class NullMetric
+   class NullTimedExecution
+     def stop
+     end
+   end
+
+   NULL_TIMED_EXECUTION_INSTANCE = NullTimedExecution.new
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/null_metric.rb]: I don't understand why do you use constans here for this kind of values? what is your way of thinking here? thanks!

File: logstash-core/lib/logstash/pipeline.rb [hunk 7]
```diff
@@ -135,13 +146,13 @@ def filters?
...
 
   def run
+    @started_at = Time.now
+
     LogStash::Util.set_thread_name("[#{pipeline_id}]-pipeline-manager")
     @logger.terminal(LogStash::Util::DefaultsPrinter.print(@settings))
 
     start_workers
 
-    running = true
```
Comment (purbon) [inline logstash-core/lib/logstash/pipeline.rb]: good catch, :+1:

File: logstash-core/lib/logstash/inputs/metrics.rb [hunk 8]
```diff
@@ -44,9 +44,9 @@ def update(snapshot)
       #   - We can use a synchronization mechanism between the called thread (update method)
       #   and the plugin thread (run method)
       #   - How we handle back pressure here?
-      #   - one snashot should be only one event
-      snapshot.metric_store.to_events.each do |event|
-        @queue << event
+      #   - one snashot should be only one event ?
+      snapshot.metric_store.all_metrics.each do |metric|
```
Comment (purbon) [inline logstash-core/lib/logstash/inputs/metrics.rb]: As I shared in the metrics store, I'm somehow in the fence for methods like this one. What do you think about a notation like

```
snapshot.metric_store.each do |metric|
   ...
end
```

File: logstash-core/lib/logstash/instrument/metric.rb [hunk 9]
```diff
@@ -10,6 +10,28 @@ class MetricNoBlockProvided < MetricException; end
   class MetricNoNamespaceProvided < MetricException; end
 
   class Metric
+    class TimedExecution
+      MILLISECONDS = 1_000_000.0
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: can we freeze this constant?

File: logstash-core/lib/logstash/instrument/metric.rb [hunk 10]
```diff
@@ -10,6 +10,28 @@ class MetricNoBlockProvided < MetricException; end
...
+
+      def start
+        @start_time = current_time
+      end
+
+      def stop
+        @metric.report_time(@key, (MILLISECONDS * (current_time - @start_time)).to_i)
+      end
+
+      def current_time
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: what is the need/purpose of `this ``current_time``` method?

Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: I would make this method private as is only internal.

File: logstash-core/lib/logstash/instrument/metric_store.rb [hunk 11]
```diff
@@ -58,13 +60,13 @@ def get_recursively(key_paths, map)
       end
     end
 
-    def to_events_recursively(values)
+    def all_metrics_recursively(values)
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: same thing for this method, naming looks confusing to me.

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 4 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: +1

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 5 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: +1

File: logstash-core/lib/logstash/instrument/null_metric.rb (hunk 6 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/null_metric.rb]: I don't want the overhead of creating  a new instance on each call since it will return a noop object.

File: logstash-core/lib/logstash/inputs/metrics.rb (hunk 8 above)
Comment (ph) [inline logstash-core/lib/logstash/inputs/metrics.rb]: I am bit worried to use `each` in this case, since I don't want people thing its an enumerable when its not. I am OK to use `.all` like you suggested earlier?

Comment (ph) [inline logstash-core/lib/logstash/inputs/metrics.rb]: For searching I was thinking more about a `#query` or a `fetch` method

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 9 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric.rb]: we can for sure.

File: logstash-core/lib/logstash/instrument/null_metric.rb (hunk 6 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/null_metric.rb]: I understand. why not having a singleton? just curious.... :smile:

Comment (ph) [inline logstash-core/lib/logstash/instrument/null_metric.rb]: both work I guess ?

Comment (purbon) [inline logstash-core/lib/logstash/instrument/null_metric.rb]: Yes, probably a singleton will be more clear. Change if we feel to do it.

**Comment by ph** [comment]
Changed `all_` for `each` and rebased with feature/metrics.

File: logstash-core/lib/logstash/filter_delegator.rb [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    def initialize(logger, klass, metric, *args)
+      options = args.reduce({}, :merge)
+
+      @logger = logger
+      @klass = klass
+      @filter = klass.new(options)
+
+      # Scope the metrics to the plugin
+      @metric = metric.namespace(@filter.identifier_name)
+      @filter.metric = @metric.namespace(:internal)
```
Comment (purbon) [inline logstash-core/lib/logstash/filter_delegator.rb]: should we update the ```internal``` namespace here? I mean as discussed in #4492.

File: logstash-core/lib/logstash/inputs/metrics.rb (hunk 8 above)
Comment (purbon) [inline logstash-core/lib/logstash/inputs/metrics.rb]: works for me!

File: logstash-core/lib/logstash/instrument/metric.rb [hunk 13]
```diff
@@ -33,19 +56,23 @@ def gauge(key, value)
...
-        content = block.call
-        duration = Time.now - start_time
-        gauge(key, duration)
+        timer = TimedExecution.new(self, key)
+        content = yield
+        timer.stop
         return content
       else
-        raise MetricNoBlockProvided
+        TimedExecution.new(self, key)
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: I understand the ``ìf``` part execute the block passed, but not sure I see here if this is executing anything, I'm I missing anything?

File: logstash-core/lib/logstash/output_delegator.rb [hunk 14]
```diff
@@ -12,11 +12,18 @@ module LogStash class OutputDelegator
...
     @threadsafe = klass.threadsafe?
     @config = args.reduce({}, :merge)
     @klass = klass
+    
+    # Create an instance of the input so we can fetch the identifier
+    output = @klass.new(*args)
+    
+    # Scope the metrics to the plugin
+    @metric = metric.namespace(output.identifier_name)
+    internal = @metric.namespace(:internal)
```
Comment (purbon) [inline logstash-core/lib/logstash/output_delegator.rb]: same things here, I guess we can remove the internal namespace now.

**Comment by purbon** [comment]
@ph very small comments, mostly leftovers and I'm good with us merging this !!! good work here!

File: logstash-core/lib/logstash/output_delegator.rb (hunk 14 above)
Comment (ph) [inline logstash-core/lib/logstash/output_delegator.rb]: agree!

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 13 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric.rb]: Something you don't want to capture a block, this method support both kind of invocation.

```ruby
timer = metric.time("responsetime")
# do work
timer.stop

or 

metric.time("responsetime")  do
# do work
end
```

File: logstash-core/lib/logstash/filter_delegator.rb (hunk 12 above)
Comment (ph) [inline logstash-core/lib/logstash/filter_delegator.rb]: agree we should removed.
